File: Leetcode/array_hashing/python3/two-sum.py
from itertools import combinations
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i, j in combinations(range(len(nums)), 2):
            if nums[i] + nums[j] == target:
                return [i, j]


# improved solution
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        numMap = {}
        n = len(nums)

        for i in range(n):
            complement = target - nums[i]
            if complement in numMap:
                return [numMap[complement], i]
            numMap[nums[i]] = i

        return []  # No solution found

File: Leetcode/array_hashing/scala/two-sum.scala
object Solution {
    def twoSum(nums: Array[Int], target: Int): Array[Int] = {
      def twoSum(index: Int, previous: Map[Int, Int]): Array[Int] = {
        previous.get(target - nums(index)) match {
          case Some(previousIndex) => Array(previousIndex, index)
          case None => twoSum(index + 1, previous ++ Map(nums(index) -> index))
        }
      }
      twoSum(0, Map.empty)
    }
  }